subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultPublishArtifactSet.java

```diff
@@ -53,6 +60,51 @@ public String toString() {
         return displayName.getDisplayName();
     }
 
+    @Override
+    public Provider<PublishArtifact> register(Provider<? extends FileSystemLocation> artifactFile, Action<ConfigurablePublishArtifact> configuration) {
+        PublishArtifactProvider provider = new PublishArtifactProvider(artifactFile, configuration);
+        addLater(provider);
+        return provider;
+    }
+
+    private static class DefaultConfigurablePublishArtifact extends DefaultPublishArtifact {
+        private final Provider<? extends FileSystemLocation> artifactFile;
+
+        public DefaultConfigurablePublishArtifact(TaskResolver resolver, Provider<? extends FileSystemLocation> artifactFile) {
+            super(resolver, null, null, null, null, null, null);
+            this.artifactFile = artifactFile;
+            this.builtBy(artifactFile);
+        }
+
+        @Override
+        public File getFile() {
+            return artifactFile.get().getAsFile();
+        }
+    }
+
+    private static class PublishArtifactProvider extends AbstractReadOnlyProvider<PublishArtifact> {
+        private final Provider<? extends FileSystemLocation> artifactFile;
+        private final Action<ConfigurablePublishArtifact> configuration;
+
+        private PublishArtifactProvider(Provider<? extends FileSystemLocation> artifactFile, Action<ConfigurablePublishArtifact> configuration) {
+            this.artifactFile = artifactFile;
+            this.configuration = configuration;
+        }
+
+        @Override
+        public Class<PublishArtifact> getType() {
+            return PublishArtifact.class;
+        }
+
+        @Override
+        public PublishArtifact getOrNull() {
+            // TODO: need task resolver to make builtBy work
```

Comment:
I think calling `builtBy()` should be an error when you give us a provider for the source file. The provider should carry the producer task information with it, or something's gone wrong.

It should also be an error to configure the source file location of the artifact using `setFile()` (which is on the implementation's superclass and visible to the Groovy DSL).

It might be worth splitting out the `ConfigurablePublishArtifact` implementation used here from `DefaultPublishArtfiact`.